File: DataStructures/Source/Algorithms/Strings.m
//
//  Strings.m
//  DataStructures
//
//  Created by Mithlesh Kumar Jha on 2/9/19.
//  Copyright © 2019 Mithlesh Kumar Jha. All rights reserved.
//

#import "Strings.h"

char* licenseKeyFormatting(char* S, int K) {

    unsigned long length = strlen(S);
    unsigned long allocatedLength = length + 1 + length/K;
    char *r = (char*)malloc(sizeof(char) * allocatedLength);
    r[allocatedLength] = 0;
    int count = 0;
    long j = allocatedLength - 1;

    for (long i= length - 1; i >= 0 && j >=0; --i) {
        char n = S[i];

        if (!isalnum(n)) {
            continue;
        }

        if (islower(n)) {
            n = toupper(n);
        }

        if (count == K) {
            r[j--] = '-';
            count = 0;
        }

        r[j--] = n;
        count++;
    }

    if (j < 0) {
        return r;
    }

    long k = j + 1;
    for (int i = k; ; ++i) {
        r[i - k] = r[i];
        if (r[i] == 0)
            break;
    }

    return r;
}

@implementation Strings

+ (NSString*) formattedLicenseKey:(NSString*)license
         subsequentFragmentLength:(NSInteger)k
               insertingCharacter:(unichar)character {

    char * str = licenseKeyFormatting(license.UTF8String, (int)k);
    // Scan the String
    NSMutableString *result = [NSMutableString new];
    NSUInteger length = license.length;
    NSInteger count = 0;
    NSString *charString = [[NSString alloc] initWithCharacters:&character length:1];

    for (NSUInteger i = length - 1; i >= 0; --i) {
        unichar nextChar = [license characterAtIndex:i];
        if (!isalnum(nextChar)) {
            if (i == 0) {
                break;
            }
            continue;
        }

        if (count == k) {
            [result insertString:charString atIndex:0];
            count = 0;
        }

        if (islower(nextChar)) {
            nextChar = toupper(nextChar);
        }

        [result insertString:[NSString stringWithCharacters:&nextChar length:1] atIndex:0];
        count++;

        if (i == 0) {
            break;
        }
    }
    return result;
}

@end

File: DataStructures/Source/Algorithms/Swift/String/Strings.swift
//
//  Strings.swift
//  DataStructures
//
//  Created by Mithlesh Kumar Jha on 2/6/19.
//  Copyright © 2019 Mithlesh Kumar Jha. All rights reserved.
//

import Foundation

public class StringAlgorithm {

    static func findAllSubstringsOf(string: String?) -> [String] {
        guard let string = string, !string.isEmpty else { return [] }
        for char in string {
            print(char)
        }

        return []
    }

    static func licenseKeyFormatting(_ S: String, _ K: Int) -> String {
        let dashCharacter: Character = "-"
        var result = S.replacingOccurrences(of: String(dashCharacter), with: "")
        let count = result.count

        guard count > 0 else {
            return ""
        }
        
        var separatorsCount = count / K

        var firstSubsequenceLength = count % K
        if firstSubsequenceLength == 0 {
            firstSubsequenceLength = K
            separatorsCount -= 1
        }

        for i in 0 ..< separatorsCount {
            if (i == 0) {
                result.insert(dashCharacter, at: result.index(result.startIndex, offsetBy: firstSubsequenceLength))
            } else {
                result.insert(dashCharacter, at: result.index(result.startIndex, offsetBy: firstSubsequenceLength + i + (i * K)))
            }
        }

        return result.uppercased()
    }
}

